test/unit_test/test_general_functions.py

```diff
@@ -0,0 +1,122 @@
+                                           {'word': 'b', 'count': 2},
+                                           {'word': 'c', 'count': 3},
+                                           {'word': 'd', 'count': 4}]}
+
+    def test_merge_list(self):
+        assert merge_list([{"a": 1, "b": 2}, {"c": 3, "d": 4}]) == {
+            'a': 1, 'b': 2, 'c': 3, 'd': 4}
+
+    def test_load_stastic(self):
+        assert load_stastic(
+            "this is a string string") == {
+            "this": 1, "is": 1, "a": 1, "string": 2}
+
+    def test_matrix_to_dict(self):
+        assert matrix_to_dict([['', 'a', 'b', 'c', 'd'], [0, 1, 2, 3, 4]]) == \
+            [{'a': 1, 'b': 2, 'c': 3, 'd': 4}]
+
+    def test_dict_to_matrix(self):
+        assert dict_to_matrix(
+            [{'a': 1, 'b': 2, 'c': 3, 'd': 4}]) == (
+                [['', 'a', 'b', 'c', 'd'], [0, 1, 2, 3, 4]],
+                ['a', 'b', 'c', 'd'])
+
+
+class TestHtmlEscape:
+    def test_amp(self):
+        assert html_escape('&') == "&amp;"
+
+    def test_quot(self):
+        assert html_escape('"') == "&quot;"
+
+    def test_apos(self):
+        assert html_escape("'") == "&apos;"
+
+    def test_gt(self):
+        assert html_escape('>') == "&gt;"
+
+    def test_lt(self):
+        assert html_escape('<') == "&lt;"
+
+    def test_all(self):
+        assert html_escape('&"\'><') == '&amp;&quot;&apos;&gt;&lt;'
+        assert html_escape("<html lang='en'></html>") == '&lt;html lang=&apo' \
+                                                         's;en&apos;&gt;&lt;' \
+                                                         '/html&gt;'
+        assert html_escape('<html lang="en"></html>') == '&lt;html lang=&quo' \
+                                                         't;en&quot;&gt;&lt;' \
+                                                         '/html&gt;'
+
+
+class TestApplyFunctionExcludeTags:
+    input_str = "<tag>asdf</tag>"
+
+    def dummy_function(self, input_string):
+        return input_string + input_string
+
+    def test_one_function(self):
+        assert apply_function_exclude_tags(
+            self.input_str, [self.dummy_function]) == '<tag>asdfasdf</tag>'
+        assert apply_function_exclude_tags(
+            self.input_str, [str.upper]) == '<tag>ASDF</tag>'
+
+    def test_two_functions(self):
+        assert apply_function_exclude_tags(
+            '<tag>asdf</tag>', [str.upper, self.dummy_function]) == '<tag>' \
+                                                                    'ASDF' \
+                                                                    'ASDF' \
+                                                                    '</tag>'
+
+    def test_multiple_functions(self):
+        assert apply_function_exclude_tags(
+            '<tag>asdf</tag>', [str.upper, str.lower,
+                                self.dummy_function]) == '<tag>asdfasdf</tag>'
+
+
+class TestDecodeBytes:
```

Comment:
This function is well done.
But there are more encoding than this.

Here introduces couple of common encoding on the web: https://stackoverflow.com/questions/8509339/what-is-the-most-common-encoding-of-each-language

Test the first 3:
UTF-8    (89.2%)
ISO-8859-1   (5.0%)
Windows-1251 (1.6%)

and also `utf-8 with bom` and `utf-16 with bom`